choreolib/src/main/java/choreo/trajectory/SwerveSample.java

```diff
@@ -320,17 +320,40 @@ public boolean equals(Object obj) {
     }
 
     var other = (SwerveSample) obj;
-    return this.t == other.t
-        && this.x == other.x
-        && this.y == other.y
-        && this.heading == other.heading
-        && this.vx == other.vx
-        && this.vy == other.vy
-        && this.omega == other.omega
-        && this.ax == other.ax
-        && this.ay == other.ay
-        && this.alpha == other.alpha
-        && Arrays.equals(this.fx, other.fx)
-        && Arrays.equals(this.fy, other.fy);
+    var fxBothNull = (this.fx == null && other.fx == null);
+    var fxNeitherNull = (this.fx != null && other.fx != null);
+    var fyBothNull = (this.fy == null && other.fy == null);
+    var fyNeitherNull = (this.fy != null && other.fy != null);
+    var equal = MathUtil.isNear(this.t       , other.t, 1E-9)
+    && MathUtil.isNear(this.x       , other.x, 1E-9)
+    && MathUtil.isNear(this.y       , other.y, 1E-9)
+    && MathUtil.isNear(this.heading , other.heading, 1E-9)
+    && MathUtil.isNear(this.vx      , other.vx, 1E-9)
+    && MathUtil.isNear(this.vy      , other.vy, 1E-9)
+    && MathUtil.isNear(this.omega   , other.omega, 1E-9)
+    && MathUtil.isNear(this.ax      , other.ax, 1E-9)
+    && MathUtil.isNear(this.ay      , other.ay, 1E-9)
+    && MathUtil.isNear(this.alpha      , other.alpha, 1E-9);
+    if (!equal) return false;
+    if (!fxBothNull) {
```

Comment:
You could refactor this to use early exits instead to avoid nesting. Also, the array equals could be refactored into a utility function.